File: medbrains/apps/android/ui/src/main/kotlin/com/medbrains/ui/CarbonExtras.kt
package com.medbrains.ui

import androidx.compose.foundation.background
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.automirrored.filled.ArrowForward
import androidx.compose.material.icons.filled.CheckCircle
import androidx.compose.material.icons.filled.Error
import androidx.compose.material.icons.filled.Info
import androidx.compose.material.icons.filled.Warning
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.OutlinedButton
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.dp

/** Carbon's tag tones. The word is always there — colour never carries the meaning alone. */
enum class CarbonTone(val fill: Color, val ink: Color, val bar: Color) {
    Neutral(Carbon.ink[2], Carbon.ink[8], Carbon.ink[5]),
    Info(Carbon.blue[1], Carbon.blue[7], Carbon.blue[5]),
    Success(Carbon.mint[1], Carbon.mint[7], Carbon.mint[4]),
    Warning(Carbon.amber[0], Carbon.amber[7], Carbon.amber[2]),
    Danger(Carbon.rose[1], Carbon.rose[7], Carbon.rose[5]),
    HighAlert(Carbon.ochre[1], Carbon.ochre[8], Carbon.ochre[5]),
}

@Composable
fun CarbonTag(text: String, tone: CarbonTone = CarbonTone.Neutral, mono: Boolean = false, modifier: Modifier = Modifier) {
    Text(
        text,
        style = if (mono) CodeText.copy(fontSize = MaterialTheme.typography.labelMedium.fontSize) else MaterialTheme.typography.labelMedium,
        color = tone.ink,
        maxLines = 1,
        modifier = modifier.background(tone.fill).padding(horizontal = 8.dp, vertical = 3.dp),
    )
}

enum class NotificationKind(val tone: CarbonTone) { Info(CarbonTone.Info), Success(CarbonTone.Success), Warning(CarbonTone.Warning), Error(CarbonTone.Danger) }

/** Carbon's inline notification: Layer-01 tile, 3dp bar, icon, title, message, optional actions. */
@Composable
fun CarbonNotification(kind: NotificationKind, title: String, message: String, modifier: Modifier = Modifier, actions: (@Composable () -> Unit)? = null) {
    val icon = when (kind) {
        NotificationKind.Info -> Icons.Filled.Info
        NotificationKind.Success -> Icons.Filled.CheckCircle
        NotificationKind.Warning -> Icons.Filled.Warning
        NotificationKind.Error -> Icons.Filled.Error
    }
    Row(modifier.fillMaxWidth().height(IntrinsicSize.Min).background(Carbon.ink[1]).semantics(mergeDescendants = true) { contentDescription = "$title. $message" }) {
        Box(Modifier.width(3.dp).fillMaxHeight().background(kind.tone.bar))
        Row(Modifier.padding(16.dp), verticalAlignment = Alignment.Top) {
            Icon(icon, contentDescription = null, tint = kind.tone.bar)
            Spacer(Modifier.width(12.dp))
            Column {
                Text(title, style = MaterialTheme.typography.titleSmall)
                Text(message, style = MaterialTheme.typography.bodyMedium)
                actions?.invoke()
            }
        }
    }
}

/** Carbon tile with a tone bar, for cards and rows that carry a state. */
@Composable
fun CarbonTonedTile(tone: CarbonTone, modifier: Modifier = Modifier, content: @Composable () -> Unit) {
    Row(modifier.fillMaxWidth().height(IntrinsicSize.Min).background(Carbon.ink[1])) {
        Box(Modifier.width(3.dp).fillMaxHeight().background(tone.bar))
        Column(Modifier.padding(16.dp).fillMaxWidth()) { content() }
    }
}

/** Carbon's tertiary button: outlined, Blue 60, label left. */
@Composable
fun CarbonTertiaryButton(text: String, onClick: () -> Unit, modifier: Modifier = Modifier, enabled: Boolean = true) {
    OutlinedButton(
        onClick = onClick,
        enabled = enabled,
        shape = MaterialTheme.shapes.small,
        modifier = modifier.height(TAP_TARGET_DP.dp),
    ) { Text(text, style = MaterialTheme.typography.bodyLarge) }
}

/** Carbon's danger button: Red 60 fill, for the act that cannot be undone. */
@Composable
fun CarbonDangerButton(text: String, onClick: () -> Unit, modifier: Modifier = Modifier, enabled: Boolean = true) {
    Button(
        onClick = onClick,
        enabled = enabled,
        shape = MaterialTheme.shapes.small,
        colors = ButtonDefaults.buttonColors(containerColor = MaterialTheme.colorScheme.error),
        modifier = modifier.fillMaxWidth().height(TAP_TARGET_DP.dp),
    ) {
        Row(Modifier.fillMaxWidth(), verticalAlignment = Alignment.CenterVertically) {
            Text(text, style = MaterialTheme.typography.bodyLarge)
            Spacer(Modifier.weight(1f))
            Icon(Icons.AutoMirrored.Filled.ArrowForward, contentDescription = null)
        }
    }
}

/** The page header every module screen opens with. */
@Composable
fun CarbonPageHeader(eyebrow: String, title: String, subtitle: String? = null, trailing: (@Composable () -> Unit)? = null) {
    Row(Modifier.fillMaxWidth().padding(start = 16.dp, end = 16.dp, top = 8.dp, bottom = 16.dp), verticalAlignment = Alignment.Top) {
        Column(Modifier.weight(1f)) {
            Eyebrow(eyebrow)
            Text(title, style = MaterialTheme.typography.headlineMedium, color = MaterialTheme.colorScheme.onSurface)
            if (subtitle != null) Text(subtitle, style = MaterialTheme.typography.bodyMedium, color = MaterialTheme.colorScheme.onSurfaceVariant)
        }
        trailing?.invoke()
    }
}

/** A Carbon stat tile: eyebrow, a large number (or an em dash), a caption. */
@Composable
fun CarbonStatTile(eyebrow: String, count: Int?, title: String, modifier: Modifier = Modifier) {
    Column(modifier.background(Carbon.ink[1]).padding(16.dp)) {
        Eyebrow(eyebrow)
        Text(count?.toString() ?: "—", style = MaterialTheme.typography.displaySmall, color = MaterialTheme.colorScheme.onSurface,
            modifier = Modifier.semantics { contentDescription = count?.toString() ?: "not available" })
        Text(title, style = MaterialTheme.typography.bodyMedium, color = MaterialTheme.colorScheme.onSurfaceVariant)
    }
}

/** A tappable structured-list row with a trailing slot. */
@Composable
fun CarbonActionRow(title: String, detail: String?, onClick: () -> Unit, modifier: Modifier = Modifier, trailing: (@Composable () -> Unit)? = null) {
    Row(
        modifier.fillMaxWidth().clickable(onClick = onClick).padding(horizontal = 16.dp, vertical = 12.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Column(Modifier.weight(1f)) {
            Text(title, style = MaterialTheme.typography.bodyLarge, color = MaterialTheme.colorScheme.onSurface)
            if (detail != null) Text(detail, style = MaterialTheme.typography.bodyMedium, color = MaterialTheme.colorScheme.onSurfaceVariant)
        }
        trailing?.invoke()
    }
    androidx.compose.material3.HorizontalDivider(color = MaterialTheme.colorScheme.outlineVariant)
}

File: medbrains/apps/ios/Packages/MedBrainsUI/Sources/MedBrainsUI/CarbonExtras.swift
import SwiftUI

/// Carbon's tag: a small filled label. The word is always there — colour
/// never carries the meaning alone (WCAG 2.2 SC 1.4.1).
public enum CarbonTone: Sendable {
    case neutral, info, success, warning, danger, highAlert

    var fill: Color {
        switch self {
        case .neutral: Carbon.ink[2]
        case .info: Carbon.blue[1]
        case .success: Carbon.mint[1]
        case .warning: Carbon.amber[0]
        case .danger: Carbon.rose[1]
        case .highAlert: Carbon.ochre[1]
        }
    }

    var ink: Color {
        switch self {
        case .neutral: Carbon.ink[8]
        case .info: Carbon.blue[7]
        case .success: Carbon.mint[7]
        case .warning: Carbon.amber[7]
        case .danger: Carbon.rose[7]
        case .highAlert: Carbon.ochre[8]
        }
    }

    /// The accent bar on a tile.
    public var bar: Color {
        switch self {
        case .neutral: Carbon.ink[5]
        case .info: Carbon.blue[5]
        case .success: Carbon.mint[4]
        case .warning: Carbon.amber[2]
        case .danger: Carbon.rose[5]
        case .highAlert: Carbon.ochre[5]
        }
    }
}

public struct CarbonTag: View {
    let text: String
    let tone: CarbonTone
    let mono: Bool

    public init(_ text: String, tone: CarbonTone = .neutral, mono: Bool = false) {
        self.text = text
        self.tone = tone
        self.mono = mono
    }

    public var body: some View {
        Text(text)
            .font(mono ? CarbonType.codeSmall : CarbonType.label)
            .foregroundStyle(tone.ink)
            .padding(.horizontal, 8)
            .padding(.vertical, 3)
            .background(tone.fill)
            .lineLimit(1)
    }
}

/// Carbon's inline notification: a Layer-01 tile with a 3pt bar in the
/// kind's colour, an icon, a title, the message, and optional actions.
public struct CarbonNotification<Actions: View>: View {
    public enum Kind { case info, success, warning, error }

    let kind: Kind
    let title: String
    let message: String
    @ViewBuilder let actions: () -> Actions

    public init(kind: Kind, title: String, message: String, @ViewBuilder actions: @escaping () -> Actions = { EmptyView() }) {
        self.kind = kind
        self.title = title
        self.message = message
        self.actions = actions
    }

    private var tone: CarbonTone {
        switch kind {
        case .info: .info
        case .success: .success
        case .warning: .warning
        case .error: .danger
        }
    }

    private var symbol: String {
        switch kind {
        case .info: "info.circle.fill"
        case .success: "checkmark.circle.fill"
        case .warning: "exclamationmark.triangle.fill"
        case .error: "exclamationmark.circle.fill"
        }
    }

    public var body: some View {
        HStack(alignment: .top, spacing: 12) {
            Image(systemName: symbol).foregroundStyle(tone.bar).padding(.top, 2)
            VStack(alignment: .leading, spacing: 4) {
                Text(title).font(CarbonType.heading01).foregroundStyle(MedBrainsTheme.ink)
                Text(message).font(CarbonType.bodyCompact).foregroundStyle(MedBrainsTheme.ink)
                actions()
            }
            Spacer(minLength: 0)
        }
        .padding(16)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(MedBrainsTheme.surface)
        .overlay(alignment: .leading) { Rectangle().fill(tone.bar).frame(width: 3) }
        .accessibilityElement(children: .combine)
    }
}

/// Carbon's tile with an accent bar, for cards and rows that carry a state.
public struct CarbonTile<Content: View>: View {
    let tone: CarbonTone
    @ViewBuilder let content: () -> Content

    public init(tone: CarbonTone = .neutral, @ViewBuilder content: @escaping () -> Content) {
        self.tone = tone
        self.content = content
    }

    public var body: some View {
        VStack(alignment: .leading, spacing: 8) { content() }
            .padding(16)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(MedBrainsTheme.surface)
            .overlay(alignment: .leading) { Rectangle().fill(tone.bar).frame(width: 3) }
    }
}

/// Carbon's tertiary button: outlined, Blue 60, for the secondary act beside a primary.
public struct CarbonTertiaryButtonStyle: ButtonStyle {
    @Environment(\.isEnabled) private var isEnabled
    public init() {}
    public func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(CarbonType.body)
            .foregroundStyle(isEnabled ? (configuration.isPressed ? MedBrainsTheme.inkInverse : MedBrainsTheme.interactive) : MedBrainsTheme.inkDisabled)
            .frame(maxWidth: .infinity, minHeight: 48, alignment: .leading)
            .padding(.horizontal, 16)
            .background(configuration.isPressed ? MedBrainsTheme.interactive : .clear)
            .overlay(Rectangle().stroke(isEnabled ? MedBrainsTheme.interactive : MedBrainsTheme.inkDisabled, lineWidth: 1))
            .contentShape(Rectangle())
    }
}

/// Carbon's danger button: Red 60 fill, for the act that cannot be undone.
public struct CarbonDangerButtonStyle: ButtonStyle {
    @Environment(\.isEnabled) private var isEnabled
    public init() {}
    public func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(CarbonType.body)
            .foregroundStyle(isEnabled ? MedBrainsTheme.inkInverse : MedBrainsTheme.inkDisabled)
            .frame(maxWidth: .infinity, minHeight: 48, alignment: .leading)
            .padding(.horizontal, 16)
            .background(isEnabled ? (configuration.isPressed ? Carbon.rose[6] : MedBrainsTheme.danger) : MedBrainsTheme.field)
            .contentShape(Rectangle())
    }
}

public extension ButtonStyle where Self == CarbonTertiaryButtonStyle {
    static var carbonTertiary: CarbonTertiaryButtonStyle { CarbonTertiaryButtonStyle() }
}

public extension ButtonStyle where Self == CarbonDangerButtonStyle {
    static var carbonDanger: CarbonDangerButtonStyle { CarbonDangerButtonStyle() }
}

/// The page header every module screen opens with: eyebrow, heading, one
/// line under it, an optional tag on the right.
public struct CarbonPageHeader<Trailing: View>: View {
    let eyebrow: String
    let title: String
    let subtitle: String?
    @ViewBuilder let trailing: () -> Trailing

    public init(eyebrow: String, title: String, subtitle: String? = nil, @ViewBuilder trailing: @escaping () -> Trailing = { EmptyView() }) {
        self.eyebrow = eyebrow
        self.title = title
        self.subtitle = subtitle
        self.trailing = trailing
    }

    public var body: some View {
        HStack(alignment: .top, spacing: 12) {
            VStack(alignment: .leading, spacing: 4) {
                Eyebrow(eyebrow)
                Text(title).font(CarbonType.heading04).foregroundStyle(MedBrainsTheme.ink)
                if let subtitle {
                    Text(subtitle).font(CarbonType.bodyCompact).foregroundStyle(MedBrainsTheme.inkSecondary)
                }
            }
            Spacer(minLength: 0)
            trailing()
        }
        .padding(.horizontal, 16)
        .padding(.top, 8)
        .padding(.bottom, 16)
    }
}
